src/io/dirops.c

```diff
@@ -13,55 +13,25 @@
 #  define PATH_MAX 2048
 #endif
 
-#ifdef _WIN32
-static wchar_t * UTF8ToUnicode(char *str)
-{
-     const int         len = MultiByteToWideChar(CP_UTF8, 0, str, -1, NULL, 0);
-     wchar_t * const result = (wchar_t *)MVM_malloc(len * sizeof(wchar_t));
-
-     MultiByteToWideChar(CP_UTF8, 0, str, -1, result, len);
-
-     return result;
-}
-
-static char * UnicodeToUTF8(const wchar_t *str)
-{
-     const int       len = WideCharToMultiByte(CP_UTF8, 0, str, -1, NULL, 0, NULL, NULL);
-     char * const result = (char *)MVM_malloc(len * sizeof(char));
-
-     WideCharToMultiByte(CP_UTF8, 0, str, -1, result, len, NULL, NULL);
-
-     return result;
-}
-
-static int mkdir_p(MVMThreadContext *tc, wchar_t *pathname, MVMint64 mode) {
-    wchar_t *p = pathname, ch;
-#else
 static int mkdir_p(MVMThreadContext *tc, char *pathname, MVMint64 mode) {
     char *p = pathname, ch;
     uv_fs_t req;
     int mkdir_error = 0;
-#endif
     int created = 0;
 
     for (;; ++p)
         if (!*p || IS_SLASH(*p)) {
             ch = *p;
             *p  = '\0';
-#ifdef _WIN32
-            created = CreateDirectoryW(pathname, NULL);
-#else
-            created = (mkdir(pathname, mode) == 0
-                       || ( (mkdir_error = errno) == EEXIST
-                       && uv_fs_stat(NULL, &req, pathname, NULL) == 0
-                       && S_ISDIR(req.statbuf.st_mode)));
-#endif
+            created = (mkdir_error = uv_fs_mkdir(NULL, &req, pathname, mode, NULL) == 0
+                       || (mkdir_error == EEXIST
+                           && uv_fs_stat(NULL, &req, pathname, NULL) == 0
+                           && S_ISDIR(req.statbuf.st_mode)));
             if (!(*p = ch)) break;
         }
 
-#ifndef _WIN32
-    errno = mkdir_error; /* error reporting should use errno of latest mkdir */
-#endif
+    uv_fs_req_cleanup(&req);
+
     if (!created) return -1;
```

Comment:
Currently this `-1` gets passed to `strerror(...)` when `mkdir_p` is called later on, which says "Unknown error". I think it needs to be something like:
```diff
diff --git a/src/io/dirops.c b/src/io/dirops.c
index fba7112fe..e97c6ddb0 100644
--- a/src/io/dirops.c
+++ b/src/io/dirops.c
@@ -32,7 +32,7 @@ static int mkdir_p(MVMThreadContext *tc, char *pathname, MVMint64 mode) {

     uv_fs_req_cleanup(&req);

-    if (!created) return -1;
+    if (!created) return mkdir_error;

     return 0;
 }
@@ -44,7 +44,7 @@ void MVM_dir_mkdir(MVMThreadContext *tc, MVMString *path, MVMint64 mode) {

     if ((mkdir_error = mkdir_p(tc, pathname, mode)) != 0) {
         MVM_free(pathname);
-        MVM_exception_throw_adhoc(tc, "Failed to mkdir: %s", strerror(mkdir_error));
+        MVM_exception_throw_adhoc(tc, "Failed to mkdir: %s", uv_strerror(mkdir_error));
     }

     MVM_free(pathname);
```

Reply:
Alternatively it could probably just `errno = mkdir_error`

Reply:
I think I changed it to `if (!created) return mkdir_error;` in an earlier version of this work, but that caused a spectest fail. I'll admit I didn't investigate whether it would have made sense to modify NQP and/or Rakudo to get the spectest to pass.

Reply:
> Currently this `-1` gets passed to `strerror(...)` when `mkdir_p` is called later on, which says "Unknown error". I think it needs to be something like:
> 
> ```diff
> diff --git a/src/io/dirops.c b/src/io/dirops.c
> index fba7112fe..e97c6ddb0 100644
> --- a/src/io/dirops.c
> +++ b/src/io/dirops.c
> @@ -32,7 +32,7 @@ static int mkdir_p(MVMThreadContext *tc, char *pathname, MVMint64 mode) {
> 
>      uv_fs_req_cleanup(&req);
> 
> -    if (!created) return -1;
> +    if (!created) return mkdir_error;
> 
>      return 0;
>  }
> @@ -44,7 +44,7 @@ void MVM_dir_mkdir(MVMThreadContext *tc, MVMString *path, MVMint64 mode) {
> 
>      if ((mkdir_error = mkdir_p(tc, pathname, mode)) != 0) {
>          MVM_free(pathname);
> -        MVM_exception_throw_adhoc(tc, "Failed to mkdir: %s", strerror(mkdir_error));
> +        MVM_exception_throw_adhoc(tc, "Failed to mkdir: %s", uv_strerror(mkdir_error));
>      }
> 
>      MVM_free(pathname);
> ```

If you change it to return `mkdir_error` and then `uv_strerror(mkdir_error)`, what does it print? Of maybe just a `fprintf(stderr, "mkdir_error = %d, error string = '%s'\n", mkdir_error, uv_strerror(mkdir_error));` in the loop after the `uv_fs_mkdir()` call would be helpful.

Reply:
I had edited that post; it since includes the error message: `No such file or directory`